### PR TITLE
Set Mode with Tests

### PR DESCRIPTION
**Pull request by chapman20j**
# What does this PR do?
Implements the set_mode function based on graph.recursive_map
Implements set_mode as a class method in MultiheadAttention, Batchnorm, and Dropout
Implements set_train and set_eval based on set_mode
Adds tests for set_mode

Fixes # (issue)

## Checklist
- [ ] This PR fixes a minor issue (e.g.: typo or small bug) or improves the docs (you can dismiss the other checks if that's the case).
- [ ] This change is discussed in a Github issue/[discussion](https://github.com/google/flax/discussions) (please add a link).
- [ x] The documentation and docstrings adhere to the [documentation guidelines](https://github.com/google/flax/blob/main/docs/README.md#how-to-write-code-documentation).
- [ x] This change includes necessary high-coverage tests. (No quality testing = no merge!)

### CODE REVIEW

File: flax/nnx/nn/attention.py [hunk 1]
```diff
@@ -638,6 +638,58 @@ def init_cache(self, input_shape: Shape, dtype: Dtype = jnp.float32):
     self.cached_value = nnx.Cache(jnp.zeros(cache_shape, dtype))
     self.cache_index = nnx.Cache(jnp.array(0, dtype=jnp.int32))
 
+  def init_cache2(
```
Comment (cgarciae) [inline flax/nnx/nn/attention.py]: We probably need to pick a better name than `init_cache2`, originally only used it for testing.

Comment (cgarciae) [reply flax/nnx/nn/attention.py]: We could also just inline this in `set_mode`. The original idea for a separate method was to try to reuse it in `__init__`.

Comment (chapman20j) [reply flax/nnx/nn/attention.py]: Just inlined the function in set_mode. Thanks for the feedback!
